src/Cardano/Wallet.hs

```diff
@@ -0,0 +1,240 @@
+        -- want to get that as an argument or, leave that decision to the caller
+        -- though it is not trivial at all. If it shrinks, it's okay because we
+        -- have enough checkpoints, but if it does increase, then we have
+        -- problems in case of rollbacks.
+        (cp' :| cp : take 2160 checkpoints)
+
+
+-- | Available balance = 'balance' . 'availableUTxO'
+availableBalance :: Wallet s -> Integer
+availableBalance =
+    balance . availableUTxO
+
+
+-- | Total balance = 'balance' . 'totalUTxO'
+totalBalance :: Wallet s -> Integer
+totalBalance =
+    balance . totalUTxO
+
+
+-- | Available UTxO = UTxO that aren't part of pending txs
+availableUTxO :: Wallet s -> UTxO
+availableUTxO (Wallet utxo pending _) =
+    utxo `excluding` txIns pending
+
+
+-- | Total UTxO = 'availableUTxO' <> "pending UTxO"
+totalUTxO :: Wallet s -> UTxO
+totalUTxO wallet@(Wallet _ pending s) =
+    let
+        -- NOTE
+        -- We _safely_ discard the state here because we aren't intending to
+        -- discover any new addresses through this operation. In practice, we
+        -- can only discover new addresses when applying blocks.
+        discardState = fst
+    in
+        availableUTxO wallet <> discardState (changeUTxO pending s)
+
+
+-- * Helpers
+
+-- | Check whether an invariants holds or not.
+--
+-- >>> invariant "not empty" [1,2,3] (not . null)
+-- [1, 2, 3]
+--
+-- >>> invariant "not empty" [] (not . null)
+-- *** Exception: not empty
+invariant
+    :: String -- ^ A title / message to throw in case of violation
+    -> a
+    -> (a -> Bool)
+    -> a
+invariant msg a predicate =
+    if predicate a then a else error msg
+
+-- | Return all transaction outputs that are ours. This plays well within a
+-- 'State' monad.
+--
+-- @
+-- myFunction :: Block -> State s Result
+-- myFunction b = do
+--    ours <- state $ txOutsOurs (transaction b)
+--    return $ someComputation ours
+-- @
+txOutsOurs
+    :: forall s. (IsOurs s)
+    => Set Tx
+    -> s
+    -> (Set TxOut, s)
+txOutsOurs txs =
+    runState $ Set.fromList <$> forMaybe (foldMap outputs txs) pick
+  where
+    pick :: TxOut -> State s (Maybe TxOut)
+    pick out = do
+        predicate <- state $ isOurs (address out)
+        return $ if predicate then Just out else Nothing
+
+    forMaybe :: Monad m => [a] -> (a -> m (Maybe b)) -> m [b]
+    forMaybe xs = fmap catMaybes . for xs
+
+-- | Construct a UTxO corresponding to a given transaction. It is important for
+-- the transaction outputs to be ordered correctly, since they become available
+-- inputs for the subsequent blocks.
+utxoFromTx :: Tx -> UTxO
+utxoFromTx tx@(Tx _ outs) =
+    UTxO $ Map.fromList $ zip (TxIn (txId tx) <$> [0..]) outs
+
+
+-- * Internals
+
+prefilterBlock
+    :: Block
+    -> Wallet s
+    -> (UTxO, Set TxIn, s)
+prefilterBlock b (Wallet !utxo _ !s) =
+    let
+        txs = transactions b
+        (ourOuts, s') = txOutsOurs txs s
```

Comment:
in the previous code prefiltering acted on the whole set of transactions. now we make sure we intersect only on those belonging to us. Nice!